Add support for mnemonics

mirceanis opened the pull request:
### What's here
This adds support for mnemonic phrases ( in reference to #3 ).
The chain of generators is as follows:

```
                   BIP39
Mnemonic phrase  <=======>  EntropyBytes
 |
 | PBKDF2
 v
Extended master key  ==> Keypair
 |
 | BIP32 derivation
 v
Extended child key   ==> Keypair

```

To create a mnemonic phrase from scratch:
`val phrase = Mnemonic.generateMnemonic()`

To generate keys from it:
`val masterEKey = Mnemonic.mnemonicToKey(phrase, "m/")`
or use a BIP44 path:
`val accountEKey = Mnemonic.mnemonicToKey(phrase, "m/44'/60'/0'/0")`

And to do something with the key:
`val signature = signMessage(message, accountEKey.getKeyPair())`

### Testing
Added test vectors from the bip32 and 39 specs.
`./gradlew :mnemonics:test` should go through them all

### Note
I had to make a few minor adjustments to other modules.
* I relaxed the BIP44 fromPath checks to allow for master path only, specified as `"m/"`
* I added some helper functions to `crypto` to handle conversion of public keys between compressed and uncompressed form
* I added some wrappers for `sha256` and `ripemd160` in `keccak_shortcut`
These can be relocated but that module seemed to be the one dealing with hashing.
* This PR only supports the ENGLISH wordlist from https://github.com/bitcoin/bips/blob/master/bip-0039/bip-0039-wordlists.md
Support for other wordlists can be the subject of another PR

ligi commented:
Hey thanks - looks good on a first glance - will have a deeper look soon

ligi commented on mnemonics/build.gradle:
```diff
@@ -0,0 +1,5 @@
+dependencies {
+    implementation fileTree(dir: 'libs', include: ['*.jar'])
```
please remove this line

mirceanis replied:
done

ligi commented on mnemonics/src/main/java/org/kethereum/bip32/BIP32.kt:
```diff
@@ -0,0 +1,18 @@
+
+import org.kethereum.bip44.BIP44
+
+object BIP32 {
```
I think we can remove this top level object - this is not java where we must do it - we have top level functions in Kotlin FTW ;-)

mirceanis replied:
done.

ligi commented on mnemonics/src/main/java/org/kethereum/bip32/ExtendedKey.kt:
```diff
@@ -0,0 +1,257 @@
+                       private val sequence: Int) {
+
+    init {
+        Security.addProvider(BouncyCastleProvider())
```
Imho this is not needed as we depend on crypto

mirceanis replied:
I cleared this as well.
You are right.

ligi commented on mnemonics/src/main/java/org/kethereum/bip32/ExtendedKey.kt:
```diff
@@ -0,0 +1,257 @@
+    /**
+     * expose keypair
+     */
+    fun getKeyPair(): ECKeyPair {
```
lets do  fun `getKeyPair() = keyPair`

ligi commented on mnemonics/src/main/java/org/kethereum/bip32/ExtendedKey.kt:
```diff
@@ -0,0 +1,257 @@
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+
+        other as ExtendedKey
```
this looks a bit odd to me

mirceanis replied:
This is the auto-generated implementation of equals and hashCode.
It's needed because of the chainCode member that is a ByteArray. It seems the hidden implementation for data classes does not compare ByteArray by contents.

An alternative would be to coerce the chainCode into a BigNumber

ligi commented on mnemonics/src/main/java/org/kethereum/bip32/ExtendedKey.kt:
```diff
@@ -0,0 +1,257 @@
+        }
+
+        private val HARDENING_FLAG = 0x80000000.toInt()
+        private fun isHardened(element: Int): Boolean {
```
this is WET as this already is in ExtendedKey - lets make it DRY

mirceanis replied:
moved the relevant code to the BIP44 module where the hardening flag already existed.

ligi commented on mnemonics/src/main/java/org/kethereum/bip32/ExtendedKey.kt:
```diff
@@ -0,0 +1,257 @@
+            return ExtendedKey(keyPair, chainCode, depth, parent, sequence)
+        }
+
+        private val EXTENDED_KEY_SIZE: Int = 78
```
think it should be const val

ligi commented on mnemonics/src/main/java/org/kethereum/bip39/Mnemonic.kt:
```diff
@@ -0,0 +1,175 @@
+object Mnemonic {
+
+    init {
+        Security.addProvider(BouncyCastleProvider())
```
think we can remove it

ligi commented on mnemonics/src/main/java/org/kethereum/bip39/Mnemonic.kt:
```diff
@@ -0,0 +1,175 @@
+        return BIP32.generateKey(generatedSeed, path)
+    }
+
+    private fun bytesToBits(data: ByteArray): BooleanArray {
```
think bytesToBits and bitsToBytes should not live here but in e.g. the extensions module

mirceanis replied:
I moved them and converted to extension functions.

mirceanis commented:
Thanks for the review so far.
I'll get back with a do-over soon

fmrsabino commented:
From a quick look it looks good! Right now we are abstracting the way wordlists are provided:  when we are generating a mnemonic phrase we can provide the `language` which will retrieve the words from a "provider" (provider has a mapping of `language -> list of words`) which depending on the implementation can retrieve from storage, network, etc...

I know that this PR only supports English for now but you can take this as a suggestion for a later PR 😉

mirceanis commented on mnemonics/src/test/java/org/kethereum/bip39/MnemonicTest.kt:
```diff
@@ -0,0 +1,251 @@
+    @Test
+    fun generatesValidPhrases() {
+
+        (1..100).forEach {
```
this may be overkill but it helped narrow down a defect in the wordlist

mirceanis commented:
I've addressed all the issues mentioned and added some extra testing that pinpointed some mistakes I had made.

What remains unsolved in this PR:
* multiple languages for wordlists (as mentioned by @fmrsabino )
* wordlist permanently in memory (as mentioned by @ligi )
* auto-generated boilerplate `equals` and `hashCode` in `ExtendedKey`

I think these can be the subject of different PRs

mirceanis commented:
Also, I just realised, something is not mentioned and may not be immediately obvious.

Any phrase (in fact any string) can generate keys. There is no verification made to ensure a one to one correspondence with an entropy buffer.
This means it is either up to the integrating app to warn the user about problems with the phrase, or this is a bug and perhaps should be addressed by another PR.

What do you think?

ligi commented:
:+1: thanks so much - much appreciated PR! Also great how fast you fixed the nits - good job all around! Will create follow-up tickets for the "What remains unsolved in this PR:" - if you are about to do them please assign yourself to the issue. Might do some of this myself also - assigning prevents collisions of effort. Think about adding this as a feature to WallETH soon and want some of the things to be cleaned up before.
